From 129dc8542db4ea4a44a27de51bd5863c7901c5f8 Mon Sep 17 00:00:00 2001
From: Quallenauge <Hamsi2k@freenet.de>
Date: Sun, 27 Feb 2022 15:27:20 +0400
Subject: [PATCH 1/2] Add support of A12 keymaster_key_blob files structure for
 legacy km

In A12 keymaster_key_blob format changed
Compared to A11 it contains another new 8 bytes at beginning "pKMblob\0" (in hex 0x704B4D626C6F6200)
We can just ignore them.

Add support for keymaster0 / keymaster1 too.

Change-Id: I6982879248ab805c7b6d8820d5b67c772f530ee4
Signed-off-by: Jabiyeff <cebiyevanar@gmail.com>
---
 crypto/fde/cryptfs.cpp | 15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)

diff --git a/crypto/fde/cryptfs.cpp b/crypto/fde/cryptfs.cpp
index 4b3cd4f9..8989533f 100644
--- a/crypto/fde/cryptfs.cpp
+++ b/crypto/fde/cryptfs.cpp
@@ -109,6 +109,7 @@ constexpr size_t INTERMEDIATE_BUF_SIZE =
 static_assert(INTERMEDIATE_BUF_SIZE == SCRYPT_LEN,
               "Mismatch of intermediate key sizes");
 
+static const std::string kPkmBlob("pKMblob\x00", 8);
 #define KEY_IN_FOOTER  "footer"
 
 #define DEFAULT_HEX_PASSWORD "64656661756c745f70617373776f7264"
@@ -464,6 +465,14 @@ static int keymaster_sign_object(struct crypt_mnt_ftr *ftr,
 
     int rc = -1;
 
+    // In A12 keymaster_key_blob format changed:
+    // it have useless for us bytes in beginning, so remove them to correctly handle key
+    std::string kmKey( &ftr->keymaster_blob[0], &ftr->keymaster_blob[0]+ftr->keymaster_blob_size );
+    if (!kmKey.compare(0, kPkmBlob.size(), kPkmBlob)) {
+        kmKey.erase(0, kPkmBlob.size());
+    }
+    keymaster_key_blob_t key = { reinterpret_cast<const uint8_t*>(kmKey.c_str()), kmKey.size() };
+
 #if TW_KEYMASTER_MAX_API >= 1
     keymaster0_device_t *keymaster0_dev = 0;
     keymaster1_device_t *keymaster1_dev = 0;
@@ -487,8 +496,8 @@ static int keymaster_sign_object(struct crypt_mnt_ftr *ftr,
 
         rc = keymaster0_dev->sign_data(keymaster0_dev,
                                       &params,
-                                      ftr->keymaster_blob,
-                                      ftr->keymaster_blob_size,
+                                      key.key_material,
+                                      key.key_material_size,
                                       to_sign,
                                       to_sign_size,
                                       signature,
@@ -497,7 +506,6 @@ static int keymaster_sign_object(struct crypt_mnt_ftr *ftr,
     }
 #if TW_KEYMASTER_MAX_API >= 1
     else if (keymaster1_dev) {
-        keymaster_key_blob_t key = { ftr->keymaster_blob, ftr->keymaster_blob_size };
         keymaster_key_param_t params[] = {
             keymaster_param_enum(KM_TAG_PADDING, KM_PAD_NONE),
             keymaster_param_enum(KM_TAG_DIGEST, KM_DIGEST_NONE),
@@ -554,7 +562,6 @@ static int keymaster_sign_object(struct crypt_mnt_ftr *ftr,
     }
 #if TW_KEYMASTER_MAX_API == 2
     else if (keymaster2_dev) {
-        keymaster_key_blob_t key = { ftr->keymaster_blob, ftr->keymaster_blob_size };
         keymaster_key_param_t params[] = {
             keymaster_param_enum(KM_TAG_PADDING, KM_PAD_NONE),
             keymaster_param_enum(KM_TAG_DIGEST, KM_DIGEST_NONE),
-- 
2.30.2

